worker/multiwatcher/worker.go

```diff
@@ -0,0 +1,352 @@
+		return errors.Trace(err)
+	}
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return errors.Trace(tomb.ErrDying)
+		case change := <-in:
+			w.config.Logger.Tracef("handle change: %#v", change)
+			// TODO: retry any errors here to deal with i/o timouts etc.
+			// Although to be honest, by the time we start getting i/o timeouts we
+			// have such a stressed system that nothing is working well.
+			if err := backing.Changed(w.store, change); err != nil {
+				return errors.Trace(err)
+			}
+		case req := <-w.request:
+			w.config.Logger.Tracef("handle request: %#v", req)
+			w.handle(req)
+		}
+		w.respond()
+	}
+}
+
+// Kill implements worker.Worker.Kill.
+func (w *Worker) Kill() {
+	w.tomb.Kill(nil)
+}
+
+// Wait implements worker.Worker.Wait.
+func (w *Worker) Wait() error {
+	return errors.Trace(w.tomb.Wait())
+}
+
+// handle processes a request from a Multiwatcher.
+func (w *Worker) handle(req *request) {
+	w.config.Logger.Tracef("start handle")
+	defer w.config.Logger.Tracef("finish handle")
+	if req.watcher.stopped {
+		w.config.Logger.Tracef("watcher %p is stopped", req.watcher)
+		// The watcher has previously been stopped.
+		if req.reply != nil {
+			select {
+			case req.reply <- false:
+			case <-w.tomb.Dying():
+			}
+		}
+		return
+	}
+	if req.reply == nil {
+		w.config.Logger.Tracef("request to stop watcher %p", req.watcher)
+		// This is a request to stop the watcher.
+		for req := w.waiting[req.watcher]; req != nil; req = req.next {
+			select {
+			case req.reply <- false:
+			case <-w.tomb.Dying():
+			}
+		}
+		delete(w.waiting, req.watcher)
+		req.watcher.stopped = true
+		w.store.DecReference(req.watcher.revno)
+		return
+	}
+	// Add request to head of list.
+	w.config.Logger.Tracef("add watcher %p request to waiting", req.watcher)
+	req.next = w.waiting[req.watcher]
+	w.waiting[req.watcher] = req
+}
+
+// respond responds to all outstanding requests that are satisfiable.
+func (w *Worker) respond() {
+	w.config.Logger.Tracef("start respond")
+	defer w.config.Logger.Tracef("finish respond")
+	for watcher, req := range w.waiting {
+		revno := watcher.revno
+		changes, latestRevno := w.store.ChangesSince(revno)
+		w.config.Logger.Tracef("%d changes since %d for watcher %p", len(changes), revno, watcher)
+		if len(changes) == 0 {
+			if req.noChanges != nil {
+				w.config.Logger.Tracef("sending down noChanges for watcher %p", watcher)
+				select {
+				case req.noChanges <- struct{}{}:
+				case <-w.tomb.Dying():
+					return
+				}
+
+				w.removeWaitingReq(watcher, req)
+			}
+			continue
+		}
+
+		req.changes = changes
+		watcher.revno = latestRevno
+
+		w.config.Logger.Tracef("sending changes down reply channel for watcher %p", watcher)
+		select {
+		case req.reply <- true:
+		case <-w.tomb.Dying():
+			return
+		}
+
+		w.removeWaitingReq(watcher, req)
+		w.store.AddReference(revno)
+	}
+}
+
+func (w *Worker) removeWaitingReq(watcher *Watcher, req *request) {
+	if req := req.next; req == nil {
```

Comment:
can we not shadow req please, this looks like a singly-linked list, so better naming would be helpful.

Reply:
Done.